src/Compilers/CSharp/Test/Semantic/Semantics/ReadOnlyStructsTests.cs

```diff
@@ -249,5 +249,505 @@ public static void Main()
+        {
+            var csharp = @"
+public struct S1
+{
+    public readonly int this[int i] => 42;
+}
+";
+            var comp = CreateCompilation(csharp);
+            comp.VerifyDiagnostics();
+        }
+
+        [Fact]
+        public void ReadOnlyGetExpressionIndexer()
+        {
+            var csharp = @"
+public struct S1
+{
+    public int this[int i]
+    {
+        readonly get => 42;
+        readonly set {}
+    }
+}
+";
+            var comp = CreateCompilation(csharp);
+            comp.VerifyDiagnostics();
+        }
+
+        [Fact]
+        public void ReadOnlyFieldLikeEvent()
+        {
+            var csharp = @"
+using System;
+
+public struct S1
+{
+    public readonly event Action<EventArgs> E;
+    public void M() { E?.Invoke(new EventArgs()); }
+}
+";
+            var comp = CreateCompilation(csharp);
+            comp.VerifyDiagnostics(
+                // (6,45): error CS0106: The modifier 'readonly' is not valid for this item
+                //     public readonly event Action<EventArgs> E;
+                Diagnostic(ErrorCode.ERR_BadMemberFlag, "E").WithArguments("readonly").WithLocation(6, 45));
+        }
+
+        [Fact]
+        public void ReadOnlyEventExplicitAddRemove()
+        {
+            var csharp = @"
+using System;
+
+public struct S1
+{
+    public readonly event Action<EventArgs> E
+    {
+        add {}
+        remove {}
+    }
+}
+";
+            var comp = CreateCompilation(csharp);
+            comp.VerifyDiagnostics();
+        }
+
+        [Fact]
+        public void ReadOnlyStaticEvent()
+        {
+            var csharp = @"
+using System;
+
+public struct S1
+{
+    public static readonly event Action<EventArgs> E
+    {
+        add {}
+        remove {}
+    }
+}
+";
+            var comp = CreateCompilation(csharp);
+            comp.VerifyDiagnostics(
+                // (6,52): error CS0106: The modifier 'readonly' is not valid for this item
+                //     public static readonly event Action<EventArgs> E
+                Diagnostic(ErrorCode.ERR_BadMemberFlag, "E").WithArguments("readonly").WithLocation(6, 52));
+        }
+
+        [Fact]
+        public void ReadOnlyEventReadOnlyAccessors()
+        {
+            var csharp = @"
+using System;
+
+public struct S1
+{
+    public event Action<EventArgs> E
+    {
+        readonly add {}
+        readonly remove {}
+    }
+}
+";
+            var comp = CreateCompilation(csharp);
+            comp.VerifyDiagnostics(
+                // (8,9): error CS1609: Modifiers cannot be placed on event accessor declarations
+                //         readonly add {}
+                Diagnostic(ErrorCode.ERR_NoModifiersOnAccessor, "readonly").WithLocation(8, 9),
+                // (9,9): error CS1609: Modifiers cannot be placed on event accessor declarations
+                //         readonly remove {}
+                Diagnostic(ErrorCode.ERR_NoModifiersOnAccessor, "readonly").WithLocation(9, 9));
+        }
```

Comment:
Should add a test or PROTOTYPE about ensuring this only works when C# 8.0 or greater is enabled.

Reply:
Done. I will probably have a PR specifically about ensuring feature checks occur appropriately in all scenarios (basically wherever the `readonly` modifier can be added).